File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.py
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.dist = x * x + y * y
        
    def __lt__(self, other):
        return self.dist > other.dist


class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        closest = []
        for x, y in points:
            heappush(closest, Point(x, y))
            
            if len(closest) > k:
                heappop(closest)
                
        return [(c.x, c.y) for c in closest]

File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.cpp
/*
min heap with comparator comparing distance from origin
*/
class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        for (auto& point : points) {
            closest.push(point);
        }
        
        vector<vector<int>> kclosest;
        while (k--) {
            kclosest.push_back(closest.top()); closest.pop();
        }
        
        return kclosest;
    }
    
private:
    class compare {
        public:
        bool operator() (vector<int>& pointa, vector<int>& pointb) {
            int dista = pointa[0] * pointa[0] + pointa[1] * pointa[1];
            int distb = pointb[0] * pointb[0] + pointb[1] * pointb[1];
            
            return dista > distb;
        }
    };
    
    priority_queue<vector<int>, vector<vector<int>>, compare> closest;
};